tests.py

```diff
@@ -1,129 +1,523 @@
+
+                LOG_IN and print(f"\n <<< received: {headers}\n")
+
+                if header_bytes == b"":
+                    break
+
+                body = self._pout.read(int(headers.get("Content-Length")))
+                LOG_IN and print(f"   < received: {body}\n")
+                self._read_q.put(header_bytes + body)
+        except Exception as ex:
+            self.exception = ex
+        self._send_q.put_nowait(None)  # stop send_loop()
+
+    # threaded
+    def _send_loop(self):
+        try:
+            while self._pin:
+                buf = self._send_q.get()
+
+                if buf is None:
+                    break
+
+                LOG_OUT and print(f"\nsending: {buf}\n")
+
+                self._pin.write(buf)
+                self._pin.flush()
+        except Exception as ex:
+            self.exception = ex
+
+    def _queue_data_to_send(self):
+        send_buf = self.lsp_client.send()
+        if send_buf:
+            self._send_q.put(send_buf)
+
+    def _read_data_received(self):
+        while not self._read_q.empty():
+            data = self._read_q.get()
+            events = self.lsp_client.recv(data)
+            for ev in events:
+                self.msgs.append(ev)
+                self._try_default_reply(ev)
+
+    def _try_default_reply(self, msg):
+        empty_reply_classes = (
+            lsp.ShowMessageRequest,
+            lsp.WorkDoneProgressCreate,
+            lsp.RegisterCapabilityRequest,
+            lsp.ConfigurationRequest,
+        )
+
+        if isinstance(msg, empty_reply_classes):
+            msg.reply()
+
+        elif isinstance(msg, lsp.WorkspaceFolders):
+            msg.reply([lsp.WorkspaceFolder(uri=self.root_uri, name="Root")])
+
+        else:
+            print(f"Cant autoreply: {type(msg)}")
+
+    def _process_qs(self):
+        self._queue_data_to_send()
+        self._read_data_received()
+
+    def get_msg_by_type(self, _type, timeout=5):
+        end_time = time.time() + timeout
+        while time.time() < end_time:
+            self._process_qs()
+
+            # raise thread's exception if have any
+            if self.exception:
+                raise self.exception
+
+            for msg in self.msgs:
+                if isinstance(msg, _type):
+                    self.msgs.remove(msg)
+                    return msg
+
+            time.sleep(0.05)
+        # end while
+
+        raise Exception(
+            f'Didn`t receive "{_type}" in time; have: {[type(m).__name__ for m in self.msgs]}'
+        )
+
+    def stop(self):
+        if self.lsp_client.is_initialized:
+            self.lsp_client.shutdown()  # send shutdown...
+            self.get_msg_by_type(lsp.Shutdown)  # receive shutdown...
+            self.lsp_client.exit()  # send exit...
+            self._process_qs()  # give data to send-thread
+        else:
+            self.process.kill()
+
+
+test_langservers = pathlib.Path(__name__).absolute().parent / "test_langservers"
```

Comment:
`__name__` is not a file name (as in `foo/bar.py`), but a module name (`foo.bar`). Maybe you meant:

```suggestion
test_langservers = pathlib.Path(__file__).absolute().parent / "test_langservers"
```

Reply:
I think that is inconsequential, since it is just a folder to install stuff.

(and that's not my code :) I just moved the line)

Reply:
Ah ok, then I have written it badly :)